spark2/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead.java

```diff
@@ -0,0 +1,388 @@
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+public class TestStructuredStreamingRead {
+  private static final Configuration CONF = new Configuration();
+  private static final Schema SCHEMA = new Schema(
+      optional(1, "id", Types.IntegerType.get()),
+      optional(2, "data", Types.StringType.get())
+  );
+  private static SparkSession spark = null;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+  @Rule
+  public ExpectedException exceptionRule = ExpectedException.none();
+
+  @BeforeClass
+  public static void startSpark() {
+    TestStructuredStreamingRead.spark = SparkSession.builder()
+        .master("local[2]")
+        .config("spark.sql.shuffle.partitions", 4)
+        .getOrCreate();
+  }
+
+  @AfterClass
+  public static void stopSpark() {
+    SparkSession currentSpark = TestStructuredStreamingRead.spark;
+    TestStructuredStreamingRead.spark = null;
+    currentSpark.stop();
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testGetChanges() throws IOException {
+    File parent = temp.newFolder("test");
+    File location = new File(parent, "table");
+    File checkpoint = new File(parent, "checkpoint");
+    Table table = createTable(location.toString());
+
+    DataSourceOptions options = new DataSourceOptions(ImmutableMap.of(
+        "path", location.toString(),
+        "checkpointLocation", checkpoint.toString()));
+    IcebergSource source = new IcebergSource();
+
+    StreamingReader streamingReader = (StreamingReader) source.createMicroBatchReader(
+        Optional.empty(), checkpoint.toString(), options);
+
+    List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+    Collections.reverse(snapshotIds);
+    long initialSnapshotId = snapshotIds.get(0);
+
+    // Getting all appends from initial snapshot.
+    List<MicroBatch> pendingBatches = streamingReader.getChangesWithRateLimit(
+        initialSnapshotId, 0, true, false, Long.MAX_VALUE);
+    Assert.assertEquals(pendingBatches.size(), 4);
+
+    List<Long> batchSnapshotIds = pendingBatches.stream()
+        .map(MicroBatch::snapshotId)
+        .collect(Collectors.toList());
+    Assert.assertEquals(batchSnapshotIds, snapshotIds);
+
+    // Getting appends from initial snapshot with index, 1st snapshot will be filtered out.
+    List<MicroBatch> pendingBatches1 = streamingReader.getChangesWithRateLimit(
+        initialSnapshotId, 1, true, false, Long.MAX_VALUE);
+
+    Assert.assertEquals(pendingBatches1.size(), 4);
+    MicroBatch batch = pendingBatches1.get(0);
+    Assert.assertEquals(batch.sizeInBytes(), 0L);
+    Assert.assertEquals(batch.endFileIndex(), 1);
+    Assert.assertTrue(Iterables.isEmpty(batch.tasks()));
```

Comment:
Why is there an empty batch? This doesn't seem right to me.

Reply:
This is to test getting batch from last index of the snapshot, since there's no more files, so it should return an empty batch.